salt/pillar/s3.py

```diff
@@ -312,6 +324,14 @@ def __get_pillar_environments(files):
         if s3_meta:
             bucket_files[bucket] = __get_pillar_files_from_s3_meta(s3_meta)
 
+            # Check if we have a NextContinuationToken and loop until we don't
+            while True:
+                continuation_token = __get_continuation_token(s3_meta)
```

Comment:
Could this be rewritten as an iterator?
```
for continuation_token in __get_continuation_tokens(s3_meta):
    ...
```

I think the function would just be:

```
for item in s3_meta:
    if item.get('NextContinuationToken'):
        yield item('NextContinuationToken')
```

Or something to that effect.

Reply:
Not sure if that approach will work.  There is only ever one `NextContinuationToken` in s3_meta, but it could be different when s3_meta is updated after subsequent calls to `__get_s3_meta`.  Once it's not longer in s3_meta then it should exist the loop since there are no more results to pull down.

Reply:
Oooooh. I see!

What about something like this?

```

    # helper s3 query function
    def __get_s3_meta():
        params = {'prefix': prefix, 'list-type': 2}
        while True:
            result = __utils__['s3.query'](
                key=creds.key,
                keyid=creds.keyid,
                kms_keyid=creds.kms_keyid,
                bucket=creds.bucket,
                service_url=creds.service_url,
                verify_ssl=creds.verify_ssl,
                location=creds.location,
                return_bin=False,
                params={'prefix': prefix},
                path_style=creds.path_style,
                https_enable=creds.https_enable)
            yield result
            continuation_token = next((item.get('NextContinuationToken')
                                       for item in s3_meta
                                       if item.get('NextContinuationToken')),
                                      None)
            if continuation_token is None:
                break
            else:
                params['continuation-token'] = continuation_token
```

Then it could simply be

```
for s3_meta in __get_s3_meta():
    ...
```

Would that work?